SafeCallerBuilderImpl fails to create proxy for bridge handlers

There are reports that the SafeCallerBuilderImpl fails creating proxy objects with the following exception:
```
2017-12-09 00:54:13.307 [ERROR] [ore.thing.internal.ThingRegistryImpl] - Could not inform the ThingTracker 'org.eclipse.smarthome.core.thing.internal.ThingManager@6138fe5f' about the 'THING_ADDED' event!
java.lang.IllegalArgumentException: repeated interface: org.eclipse.smarthome.core.thing.binding.BridgeHandler
        at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:597) [?:?]
        at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:557) [?:?]
        at java.lang.reflect.WeakCache$Factory.get(WeakCache.java:230) [?:?]
        at java.lang.reflect.WeakCache.get(WeakCache.java:127) [?:?]
        at java.lang.reflect.Proxy.getProxyClass0(Proxy.java:419) [?:?]
        at java.lang.reflect.Proxy.newProxyInstance(Proxy.java:719) [?:?]
        at org.eclipse.smarthome.core.internal.common.SafeCallerBuilderImpl.build(SafeCallerBuilderImpl.java:65) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.doInitializeHandler(ThingManager.java:630) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.initializeHandler(ThingManager.java:557) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.registerAndInitializeHandler(ThingManager.java:914) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingManager.thingAdded(ThingManager.java:370) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyTrackers(ThingRegistryImpl.java:213) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutAddedElement(ThingRegistryImpl.java:132) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.notifyListenersAboutAddedElement(ThingRegistryImpl.java:1) [116:org.eclipse.smarthome.core.thing:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:136) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractRegistry.added(AbstractRegistry.java:1) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:56) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListeners(AbstractProvider.java:74) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractProvider.notifyListenersAboutAddedElement(AbstractProvider.java:78) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractManagedProvider.add(AbstractManagedProvider.java:64) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.core.common.registry.AbstractRegistry.add(AbstractRegistry.java:210) [109:org.eclipse.smarthome.core:0.9.0.201712011551]
        at org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource.create(ThingResource.java:213) [126:org.eclipse.smarthome.io.rest.core:0.9.0.201712011551]
```
As this most likely happens for many/most bindings, I consider this a very critical bug.

--- a/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/internal/common/SafeCallerImplTest.java
+++ b/bundles/core/org.eclipse.smarthome.core.test/src/test/java/org/eclipse/smarthome/core/internal/common/SafeCallerImplTest.java
@@ -35,7 +35,6 @@ import org.eclipse.smarthome.core.common.QueueingThreadPoolExecutor;
 import org.eclipse.smarthome.test.java.JavaTest;
 import org.junit.After;
 import org.junit.Before;
-import org.junit.Ignore;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.TestName;
@@ -75,6 +74,9 @@ public class SafeCallerImplTest extends JavaTest {
         public String method() {
             return "Hello";
         }
+    }
+
+    public static class DerivedTarget extends Target implements ITarget {
     }
 
     @Before
@@ -494,6 +496,12 @@ public class SafeCallerImplTest extends JavaTest {
         for (int actual : q) {
             assertThat(actual, is(expected++));
         }
+    }
+
+    @Test
+    public void testDuplicateInterface() {
+        ITarget target = new DerivedTarget();
+        safeCaller.create(target).build().method();
     }
 
     private void assertDurationBelow(long high, Runnable runnable) {

--- a/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/common/SafeCallerImpl.java
+++ b/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/common/SafeCallerImpl.java
@@ -12,10 +12,10 @@
  */
 package org.eclipse.smarthome.core.internal.common;
 
-import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.List;
+import java.util.HashSet;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.Executors;
 import java.util.concurrent.ScheduledExecutorService;
 
@@ -86,7 +86,7 @@ public class SafeCallerImpl implements SafeCaller {
     }
 
     private static <T> Class<?>[] getAllInterfaces(T target) {
-        List<Class<?>> ret = new ArrayList<>();
+        Set<Class<?>> ret = new HashSet<>();
         Class<?> clazz = target.getClass();
         while (clazz != null) {
             ret.addAll(Arrays.asList(clazz.getInterfaces()));

